tools/ktx/utility.h

```diff
@@ -0,0 +1,241 @@
+        return comp(std::invoke(proj, lhs), std::invoke(proj, rhs));
+    });
+}
+
+template <typename Range, typename Comp = std::less<>, typename Proj = identity>
+constexpr inline void sort(Range& range, Comp&& comp = {}, Proj&& proj = {}) {
+    return std::sort(std::begin(range), std::end(range), [&](const auto& lhs, const auto& rhs) {
+        return comp(std::invoke(proj, lhs), std::invoke(proj, rhs));
+    });
+}
+
+inline void replace_all_inplace(std::string& string, std::string_view search, std::string_view replace) {
+    auto pos = string.find(search);
+    while (pos != std::string::npos) {
+        string.replace(pos, search.size(), replace);
+        pos = string.find(search, pos + replace.size());
+    }
+}
+
+[[nodiscard]] inline std::string replace_all_copy(std::string string, std::string_view search, std::string_view replace) {
+    replace_all_inplace(string, search, replace);
+    return string;
+}
+
+[[nodiscard]] inline std::string escape_json_copy(std::string string) {
+    replace_all_inplace(string, "\\", "\\\\");
+    replace_all_inplace(string, "\"", "\\\"");
+    replace_all_inplace(string, "\n", "\\n");
+    return string;
+}
+
+// --- UTF-8 ---------------------------------------------------------------------------------------
+
+/**
+ * @internal
+ * @brief Given the lead byte of a UTF-8 sequence returns the expected length of the codepoint
+ * @param[in] leadByte The lead byte of a UTF-8 sequence
+ * @return The expected length of the codepoint */
+[[nodiscard]] constexpr inline int sequenceLength(uint8_t leadByte) noexcept {
+    if ((leadByte & 0b1000'0000u) == 0b0000'0000u)
+        return 1;
+    if ((leadByte & 0b1110'0000u) == 0b1100'0000u)
+        return 2;
+    if ((leadByte & 0b1111'0000u) == 0b1110'0000u)
+        return 3;
+    if ((leadByte & 0b1111'1000u) == 0b1111'0000u)
+        return 4;
+
+    return 0;
+}
+
+/**
+ * @internal
+ * @brief Checks if the codepoint was coded as a longer than required sequence
+ * @param[in] codepoint The unicode codepoint
+ * @param[in] length The UTF-8 sequence length
+ * @return True if the sequence length was inappropriate for the given codepoint */
+[[nodiscard]] constexpr inline bool isOverlongSequence(uint32_t codepoint, int length) noexcept {
+    if (codepoint < 0x80)
+        return length != 1;
+    else if (codepoint < 0x800)
+        return length != 2;
+    else if (codepoint < 0x10000)
+        return length != 3;
+    else
+        return false;
+}
+
+/**
+ * @internal
+ * @brief Checks if the codepoint is valid
+ * @param[in] codepoint The unicode codepoint
+ * @return True if the codepoint is a valid unicode codepoint */
+[[nodiscard]] constexpr inline bool isCodepointValid(uint32_t codepoint) noexcept {
+    return codepoint <= 0x0010FFFFu
+            && !(0xD800u <= codepoint && codepoint <= 0xDBFFu);
+}
+
+/**
+ * @internal
+ * @brief Safely checks and advances an UTF-8 sequence iterator to the next unicode codepoint
```

Comment:
Should be "a UTF-8". This is one of the examples where English apparently breaks its own rules. This is because UTF-8 is pronounced with a hard vowel as in "you".